src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -134,87 +134,341 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
+                            break;
+                        case EETypeElementType.UInt64:
+                            doubleValue = (double)RuntimeHelpers.FastUnbox<ulong>(srcObject);
+                            break;
+                        case EETypeElementType.Single:
+                            doubleValue = (double)RuntimeHelpers.FastUnbox<float>(srcObject);
+                            break;
+                        case EETypeElementType.Double:
+                            doubleValue = RuntimeHelpers.FastUnbox<double>(srcObject);
+                            break;
+                        default:
+                            goto Failure;
                     }
+                    rawDstValue = &doubleValue;
                     break;
 
                 default:
-                    Debug.Fail("Unexpected CorElementType: " + dstElementType + ": Not a valid widening target.");
-                    dstObject = null;
-                    return CreateChangeTypeException(srcEEType, dstEEType, semantics);
+                    goto Failure;
+            }
+
+            if (rawDstValue != null)
+            {
+                if (dstEEType->IsEnum)
+                {
+                    long rawEnumValue;
+                    switch (dstElementType)
+                    {
+                        case EETypeElementType.Char:
+                            rawEnumValue = (long)charValue;
+                            break;
+                        case EETypeElementType.SByte:
+                            rawEnumValue = (long)sbyteValue;
+                            break;
+                        case EETypeElementType.Byte:
+                            rawEnumValue = (long)byteValue;
+                            break;
+                        case EETypeElementType.Int16:
+                            rawEnumValue = (long)shortValue;
+                            break;
+                        case EETypeElementType.UInt16:
+                            rawEnumValue = (long)ushortValue;
+                            break;
+                        case EETypeElementType.Int32:
+                            rawEnumValue = (long)intValue;
+                            break;
+                        case EETypeElementType.UInt32:
+                            rawEnumValue = (long)uintValue;
+                            break;
+                        case EETypeElementType.Int64:
+                            rawEnumValue = longValue;
+                            break;
+                        default:
+                            rawEnumValue = (long)ulongValue;
+                            break;
+                    }
+                    dstObject = Enum.ToObject(dstEEType, rawEnumValue);
+                }
+                else
+                {
+                    dstObject = RuntimeHelpers.Box(ref *(byte*)&rawDstValue, new RuntimeTypeHandle(dstEEType));
```

Comment:
This should work for all cases - both primitives and enums.

Reply:
Oh cool, didn't realize that! Fixed in eeb38bf83b10386755db2e0eb883d6c6f8f03aa7 🙂